patroni/quorum.py

```diff
@@ -0,0 +1,138 @@
+            if add_to_sync:
+                yield self.sync_update(self.numsync, self.sync | add_to_sync)
+        elif self.sync > self.voters:
+            # Case 2: sync is superset of quorum nodes. In the middle of changing replication factor.
+            # Add to quorum voters nodes that are already synced and active
+            add_to_quorum = (self.sync - self.voters) & self.active
+            if add_to_quorum:
+                yield self.quorum_update(
+                        quorum=self.quorum,
+                        voters=self.voters | add_to_quorum)
+            # Remove from sync nodes that are dead
+            remove_from_sync = self.sync - self.voters
+            if remove_from_sync:
+                yield self.sync_update(
+                        numsync=min(self.sync_wanted, len(self.sync) - len(remove_from_sync)),
+                        sync=self.sync - remove_from_sync)
+
+        # After handling these two cases quorum and sync must match.
+        assert self.voters == self.sync
+
+        safety_margin = self.quorum + self.numsync - len(self.voters|self.sync)
+        if safety_margin > 1:
+            # Case 3: quorum or replication factor is bigger than needed. In the middle of changing requested replication factor.
+            if self.numsync > self.sync_wanted:
+                # Reduce replication factor
+                yield self.sync_update(min(self.sync_wanted, len(self.sync)), self.sync)
+            elif len(self.voters) > self.numsync:
+                # Reduce quorum
+                yield self.quorum_update(len(self.voters) + 1 - self.numsync, self.voters)
+
+        # We are in a steady state point. Find if desired state is different and act accordingly.
+
+        # If any nodes have gone away, evict them
+        to_remove = self.sync - self.active
+        if to_remove:
+            can_reduce_quorum_by = self.quorum - 1
+            # If we can reduce quorum size try to do so first
+            if can_reduce_quorum_by:
+                 # Pick nodes to remove by sorted order to provide deterministic behavior for tests
+                remove = set(sorted(to_remove, reverse=True)[:can_reduce_quorum_by])
+                yield self.sync_update(self.numsync, self.sync - remove)
+                yield self.quorum_update(self.quorum - can_reduce_quorum_by, self.voters - remove)
+                to_remove &= self.sync
+            if to_remove:
+                assert self.quorum == 1
+                yield self.quorum_update(self.quorum, self.voters - to_remove)
+                yield self.sync_update(self.numsync - len(to_remove), self.sync - to_remove)
+
+        # If any new nodes, join them to quorum
+        to_add = self.active - self.sync
+        if to_add:
+            # First get to requested replication factor
+            increase_numsync_by = self.sync_wanted - self.numsync
+            if increase_numsync_by:
+                add = set(sorted(to_add)[:increase_numsync_by])
```

Comment:
This one I don't really understand.

It is responsible for https://github.com/zalando/patroni/pull/672/files#diff-57e7f41a6358e844d5245ea7b8409311R12
```python
+        self.assertEquals(list(QuorumStateResolver(1, set("a"), 1, set("a"), active=set("abcde"), sync_wanted=3)), [
+            ('sync', 3, set('abc')),
+            ('quorum', 3, set('abcde')),
+            ('sync', 3, set('abcde')),
+        ])
```

Why does it fist adds only 'b' and 'c'? Why not all 4 at the same time? Something like:
```python
+            ('sync', 3, set('abcde')),
+            ('quorum', 3, set('abcde')),
```

Reply:
Yes, that would be a correct optimization. Currently the algorithm doesn't take into account that the master is special and always guaranteed to be one of the synced nodes. This may lead to extra steps when adding multiple nodes at the same time. On the plus side this allows for promotion before we update sync state in DCS. I need to think a bit about how the algorithm would look like if we consider that master to be special. If it makes it considerably more complicated than it already is, then it might not be worth the minor improvement in efficiency. I will also write down how and why the QuorumStateResolver works.